In [1]:
from pathlib import Path
import pandas as pd

PROJECT_ROOT = Path.cwd().parent

ENCODED_FILE = (
    PROJECT_ROOT
    / "data"
    / "processed"
    / "student_data_encoded.csv"
)

df_encoded = pd.read_csv(ENCODED_FILE)

print("Encoded dataset loaded.")
print("Shape:", df_encoded.shape)

Encoded dataset loaded.
Shape: (395, 42)


In [2]:
required_grade_columns = ["G1", "G2", "G3"]

missing_grade_columns = [
    col for col in required_grade_columns
    if col not in df_encoded.columns
]

print("Missing columns:", missing_grade_columns)

Missing columns: []


In [3]:
duplicate_columns = (
    df_encoded.columns[
        df_encoded.columns.duplicated()
    ]
    .tolist()
)

print("Duplicate columns:", duplicate_columns)

Duplicate columns: []


In [4]:
X_full = df_encoded.drop(columns=["G3"]).copy()

y = df_encoded["G3"].copy()

print("Full-information features:", X_full.shape[1])

Full-information features: 41


In [5]:
print("Original encoded dataset shape:", df_encoded.shape)
print("Feature matrix shape:", X_full.shape)
print("Target vector shape:", y.shape)

Original encoded dataset shape: (395, 42)
Feature matrix shape: (395, 41)
Target vector shape: (395,)


In [6]:
assert "G3" not in X_full.columns

print("Verification passed: G3 is not present in X_full.")

Verification passed: G3 is not present in X_full.


In [7]:
assert "G1" in X_full.columns
assert "G2" in X_full.columns

print("Verification passed: G1 and G2 remain in X_full.")

Verification passed: G1 and G2 remain in X_full.


In [8]:
FULL_INFO_FILE = (
    PROJECT_ROOT
    / "data"
    / "processed"
    / "student_full_information.csv"
)

X_full.to_csv(FULL_INFO_FILE, index=False)

print("Saved:", FULL_INFO_FILE)

Saved: /Users/jjun28/Desktop/Github 2026/student-performance-prediction-ml/data/processed/student_full_information.csv


# Session 20: Feature Engineering II

## Full-Information Feature Set

### Objective

Create the full-information dataset for predictive modeling.

### Results

- Encoded dataset shape: (395, 42)
- Feature matrix shape: (395, 41)
- Target vector shape: (395,)

### Interpretation

The full-information scenario includes all available predictors, including prior academic grades G1 and G2. The final grade G3 was separated and used as the target variable.

This feature set is expected to achieve the highest predictive accuracy because it includes the strongest academic indicators available before the final grade is assigned.